Say goodbye when terminating

joerg-krause opened the pull request:
Update the dependencies and say goodbye to the Spotify Connect client when librespot is terminate so it stops playback and removes the device from its list.

Closes: #114

joerg-krause committed:
Update dependencies

plietar commented on src/lib.rs:
```diff
@@ -1,5 +1,6 @@
 #![crate_name = "librespot"]
 
+#![feature(const_fn)]
```
That would break builds on stable

plietar commented on src/player.rs:
```diff
@@ -11,6 +13,18 @@ use metadata::{FileFormat, Track, TrackRef};
+
+extern fn sighandler_exit(_: libc::c_int) {
+    unsafe {
+        received_signal.store(15, Ordering::Relaxed);
```
15 ?

joerg-krause replied:
15 is the portable number for SIGTERM.

plietar commented on src/player.rs:
```diff
@@ -11,6 +13,18 @@ use metadata::{FileFormat, Track, TrackRef};
+
+use nix::sys::signal;
+
+static mut received_signal: AtomicUsize = AtomicUsize::new(0);
```
You can use `ATOMIC_USIZE_INIT`, which would remove the need for `const_fn`.
Also why a usize rather a bool ?

joerg-krause replied:
Agreed. I will replace it. I would like to add support for handling more signals, thats why an integer.

plietar reviewed (approved):
`nix` is really too low level for this. You could use something like https://github.com/tailhook/signal.
The player shouldn't have anything to do with this. I'm generally not comfortable having anything outside of main which has a global effect.
I would do all the signal stuff in main, just call `spirc.n otify_goobye()` and exit.

joerg-krause commented:
My first try was to run it in main, but I moved it to the player loop, because this was the only option to check for a signal.

I will play with signal and chan-signal and see if I am able to use them with my limited Rust knowledge :smile:

joerg-krause commented:
@plietar I need some help!

I am trying to use [simple-signal](https://crates.io/crates/simple-signal):

``` rust
    let spirc = SpircManager::new(session.clone(), player);
    thread::spawn(move || spirc.run());
    Signals::set_handler(&[Signal::Int, Signal::Term],
        move |signals| {
            println!("Caught: {:?}", signals);
            spirc.send_goodbye();
            exit(0);
        }
    );
```

This fails with: `error[E0382]: capture of moved value: spirc`.

Unfortunately, my Rust skills are still limited, so can you give me a hint?

joerg-krause commented:
Is it as easy as to clone spirc? Like this:

``` rust
    let spirc = SpircManager::new(session.clone(), player);
    let spirc_signal = spirc.clone();
    thread::spawn(move || spirc.run());
    Signals::set_handler(&[Signal::Int, Signal::Term],
        move |signals| {
            println!("Caught: {:?}", signals);
            spirc_signal.send_goodbye();
            exit(0);
        }
    );
```

Or should it be done in a different way?

plietar commented:
Yeah that looks good. `SpircManager` is reference counted and can be cloned and shared between threads.

joerg-krause committed:
Say goodbye when terminating

When librespot is terminated while a session is active it will now send
a goodbye message, so that the Spotify client unregisters the device
from its list.

Closes: #114

joerg-krause commented:
Thanks! I've updated the commit...
